src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerGenotypingEngine.java

```diff
@@ -168,24 +273,82 @@ public CalledHaplotypes(final List<VariantContext> calls, final Set<Haplotype> c
         mergedVC.getAlternateAlleles().stream().filter(a -> !a.isSymbolic()).forEach(a -> result.put(a, new ArrayList<>()));
 
         for (final Haplotype h : haplotypes) {
-            final VariantContext haplotypeVC = h.getEventMap().get(loc);
-            if (haplotypeVC == null) {    //no events --> this haplotype supports the reference at this locus
-                result.get(ref).add(h);
-            } else {
-                //TODO: this only works if eventsAtThisLoc's and mergedVC's alleles are ordered identically.
-                //TODO: this is in fact the case (see AssemblyBasedCallerUtils::makeMergedVariantContext), but
-                //TODO: that's not good enough
-                for (int n = 0; n < eventsAtThisLoc.size(); n++) {
-                    if (haplotypeVC.hasSameAllelesAs(eventsAtThisLoc.get(n))) {
-                        result.get(mergedVC.getAlternateAllele(n)).add(h);
-                        break;
+
+            final Iterator<VariantContext> spanningEvents = h.getEventMap().getSpanningEvents(loc);
+
+            final boolean hasSpanningEvents = spanningEvents.hasNext();
+
+            while (spanningEvents.hasNext()) {
+                final VariantContext spanningEvent = spanningEvents.next();
+                if (spanningEvent.getStart() == loc) {
+                    // the event starts at the current location
+
+                    if (spanningEvent.getReference().length() == mergedVC.getReference().length()) {
+                        // reference allele lengths are equal; we can just use the spanning event's alt allele
+                        // in the case of GGA mode the spanning event might not match an allele in the mergedVC
+                        if (result.containsKey(spanningEvent.getAlternateAllele(0))) {
+                            result.get(spanningEvent.getAlternateAllele(0)).add(h);
+                        }
+                    } else if (spanningEvent.getReference().length() < mergedVC.getReference().length()) {
+                        // spanning event has shorter ref allele than merged VC; we need to pad out its alt allele
+                        final Map<Allele, Allele> spanningEventAlleleMappingToMergedVc
+                                = GATKVariantContextUtils.createAlleleMapping(mergedVC.getReference(), spanningEvent, new ArrayList<>());
+                        final Allele remappedSpanningEventAltAllele = spanningEventAlleleMappingToMergedVc.get(spanningEvent.getAlternateAllele(0));
+                        // in the case of GGA mode the spanning event might not match an allele in the mergedVC
```

Comment:
Why do we do nothing with the spanning event in such a case?

Reply:
What's in the merged VC is either the union of the alt alleles from the active haplotypes or the GGA alleles. The comment refers to the case when an assembled haplotype supports an event that isn't present in the merged VC. HaplotypeCaller's GGA mode behavior is to ignore active events that aren't in the given alleles list so my assumption was that those haplotypes should be dropped from the event map. Does that seem correct to you?

Reply:
That does sound correct, and I see that you hooked up `SomaticGenotypingEngine` to the version of this method that doesn't get the GGA alleles, which is correct because they have already been injected into the haplotypes.  So, yes, I think everything is okay.